core/valor_core/classification.py

```diff
@@ -0,0 +1,1343 @@
+import time
+from collections import defaultdict
+from typing import Dict, List, Optional, Tuple, Union
+
+import numpy as np
+import pandas as pd
+from valor_core import enums, metrics, schemas, utilities
+
+
+def _create_classification_grouper_mappings(
+    label_map: Optional[schemas.LabelMapType],
+    labels: list,
+) -> Dict[str, dict]:
+    """Create grouper mappings for use when evaluating classifications."""
+    mapping_dict = dict()
+    if label_map:
+        for label, grouper in label_map.items():
+            mapping_dict[(label.key, label.value)] = (
+                grouper.key,
+                grouper.value,
+            )
+
+    # define mappers to connect groupers with labels
+    label_value_to_grouper_value = dict()
+    label_to_grouper_key_mapping = dict()
+
+    for label_key, label_value in labels:
+        # the grouper should equal the (label.key, label.value) if it wasn't mapped by the user
+        grouper_key, grouper_value = mapping_dict.get(
+            (label_key, label_value), (label_key, label_value)
+        )
+
+        label_value_to_grouper_value[label_value] = grouper_value
+        label_to_grouper_key_mapping[(label_key, label_value)] = grouper_key
+
+    return {
+        "label_value_to_grouper_value": label_value_to_grouper_value,
+        "label_to_grouper_key_mapping": label_to_grouper_key_mapping,
+    }
+
+
+def _add_columns_to_groundtruth_and_prediction_table(
+    df: pd.DataFrame, grouper_mappings: dict
+) -> None:
+    """Add label, grouper_key, and grouper_value columns to a particular dataframe. Modifies the dataframe in place."""
+
+    df["label"] = df.apply(
+        lambda chain_df: (chain_df["label_key"], chain_df["label_value"]),
+        axis=1,
+    )
+    df["grouper_key"] = df["label"].map(
+        grouper_mappings["label_to_grouper_key_mapping"]
+    )
+    df["grouper_value"] = df["label_value"].map(
+        grouper_mappings["label_value_to_grouper_value"]
+    )
+
+
+def _calculate_confusion_matrix_df(
+    merged_groundtruths_and_predictions_df: pd.DataFrame,
+) -> tuple[pd.DataFrame, list[metrics.ConfusionMatrix]]:
+    """Calculate our confusion matrix dataframe."""
+
+    cm_counts_df = (
+        merged_groundtruths_and_predictions_df[
+            ["grouper_key", "pd_grouper_value", "gt_grouper_value"]
+        ]
+        .groupby(
+            ["grouper_key", "pd_grouper_value", "gt_grouper_value"],
+            as_index=False,
+            dropna=False,
+        )
+        .size()
+    )
+
+    cm_counts_df["true_positive_flag"] = (
+        cm_counts_df["pd_grouper_value"] == cm_counts_df["gt_grouper_value"]
+    )
+
+    # resolve pandas typing error
+    assert isinstance(cm_counts_df, pd.DataFrame)
```

Comment:
`assert` should be replaced with a if-statement that raises. 

See [reference](https://www.reddit.com/r/learnpython/comments/zd217y/is_using_assert_considered_good_practice/).

Reply:
this is probably more of a nit since this condition is for typing errors and doesn't affect runtime.

Reply:
fixed